typed_syscall_data/README.md

```diff
@@ -0,0 +1,481 @@
+| `0b1110` | *Reserved for future use*    | -               | -                | -                         |
+| `0b1111` | Arbitrary register value     | DNE             | DNE              | `Register`                |
+
+We can describe a list of N types as a 4N bit integer by embedding the Nth type
+ID in the Nth nibble of the integer. So:
+
+* The empty list has ID `0`
+* A list of one type has an ID equal to that type.
+* A list of two types has the first type in bits 0-3 and the second in bits 4-7.
+* A list of three types has the first type in bits 0-3, the second in bits 4-7,
+  and the third in bits 8-11.
+* and so on
+
+For example, the type `(bool, u32, *mut T)` would be described by
+`0b110000101010`, expanded here:
+
+| Bits | Value    | Type           |
+| ---- | -------- | -------------- |
+| 0-3  | `0b1010` | `bool`         |
+| 4-7  | `0b0010` | `u32`          |
+| 8-11 | `0b1100` | Narrow pointer |
+
+Note that if this type descriptor were stored in a larger type (such as a
+`u32`), you can determine that it is a list of three types because `0b0000` is
+not a valid type ID.
+
+## Putting multiple differently-typed values into registers
+
+This table indicates how many registers are needed for each type:
+
+| Type              | 32 bit non-CHERI | 64 bit non-CHERI | 32 bit CHERI | 64 bit CHERI |
+| ----------------- | ---------------- | ---------------- | ------------ | ------------ |
+| `u64`             | 2                | 1                | 2            | 1            |
+| `i64`             | 2                | 1                | 2            | 1            |
+| `f64`             | 2                | 1                | 2            | 1            |
+| CHERI capability  | N/A              | N/A              | 1            | 1            |
+| *Everything else* | 1                | 1                | 1            | 1            |
+
+Shorthand: we'll use `regcount<T>` to denote the number of registers needed for
+type `T` on a particular platform.
+
+If we have a list of typed values `(v1: T1, v2: T2, v3: T3, ...)` and an ordered
+list of registers, we can store `v1` in the first `regcount<T1>` registers, `v2`
+in the next `regcount<T2>` registers, `v3` in the next `regcount<T3>` registers,
+and so on. Values that span multiple registers store their least significant
+bits in the first register and their most significant bits in the second
+register. To make this concrete, if we have values `(v1: bool, v2: u64, v3: *mut
+())` we would pack them as follows on a 32 bit non-CHERI system:
+
+1. `v1`
+2. Least-significant 32 bits of `v2`
+3. Most-significant 32 bits of `v2`
+4. `v3`
+
+## ArbitraryData
+
+`ArbitraryData` is a data structure designed to fit in registers and carry data
+of a variety of types. The value of the first register is the type descriptor
+for the list of data types. The remaining pieces of data are stored in the rest
+of the registers in order (as described in the previous heading).
```

Comment:
Something I'm not clear about at this point in the document is what problem `ArbitraryData` solves. How it works totally makes sense, but why does it need to exist?

Reply:
From later, it looks like the idea is to replace Section 3.2 of TRD104 with something far more flexible and robust.

Reply:
I added a paragraph to the top of the ArbitraryData section that elaborates on its purpose and how it related to TRD 104.